base/checked.jl

```diff
@@ -0,0 +1,333 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
+
+# Support for checked integer arithmetic
+
+module Checked
+
+import Base: checked_neg, checked_abs, checked_add, checked_sub, checked_mul,
+       checked_div, checked_rem, checked_fld, checked_mod, checked_cld
+export checked_neg, checked_abs, checked_add, checked_sub, checked_mul,
+       checked_div, checked_rem, checked_fld, checked_mod, checked_cld
+
+import Core.Intrinsics: box, unbox,
+       checked_sadd_int, checked_ssub_int, checked_smul_int, checked_sdiv_int,
+       checked_srem_int,
+       checked_uadd_int, checked_usub_int, checked_umul_int, checked_udiv_int,
+       checked_urem_int
+
+typealias SignedInt Union{Int8,Int16,Int32,Int64,Int128}
+typealias UnsignedInt Union{UInt8,UInt16,UInt32,UInt64,UInt128}
+
+# LLVM has several code generation bugs for checked integer arithmetic (see e.g.
+# #4905). We thus distinguish between operations that can be implemented via
+# intrinsics, and operations for which we have to provide work-arounds.
+
+# Note: As far as this code has been tested, most checked_* functions are
+# working fine in LLVM. (Note that division is still handled via `base/int.jl`,
+# which always checks for overflow, and which provides its own sets of
+# work-arounds for LLVM codegen bugs.) However, the comments in `base/int.jl`
+# and in issue #4905 are more pessimistic. For the time being, we thus retain
+# the ability to handle codegen bugs in LLVM, until the code here has been
+# tested on more systems and architectures. It also seems that things depend on
+# which compiler that was used to build LLVM (i.e. either gcc or clang).
+
+# These unions are used for most checked functions:
+#     BrokenSignedInt
+#     BrokenUnsignedInt
+# These unions are used for checked_{mul,div,rem}:
+#     BrokenSignedIntMul
+#     BrokenUnsignedIntMul
+
+# This code runs early during bootstrap, and we can't use Julia's version
+# strings yet
+const llvm_version = Int(ccall(:jl_get_LLVM_VERSION, UInt32, ()))
+if llvm_version >= 30500
+    if WORD_SIZE == 32
+        typealias BrokenSignedInt Union{}
```

Comment:
See the comment just above. They are the same at the moment, but as we explore other combinations of system architectures and LLVM versions, we may need to tag some types as broken. I started out this way, and then pruned the set of broken types and LLVM combinations. I'm not quite sure yet that I didn't over-prune.